// simpl-schema/dist/testHelpers/friendsSchema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _SimpleSchema = require("../SimpleSchema");

const friendsSchema = new _SimpleSchema.SimpleSchema({
  name: {
    type: String,
    optional: true
  },
  friends: {
    type: Array,
    minCount: 1
  },
  'friends.$': {
    type: Object
  },
  'friends.$.name': {
    type: String,
    max: 3
  },
  'friends.$.type': {
    type: String,
    allowedValues: ['best', 'good', 'bad']
  },
  'friends.$.a': {
    type: Object,
    optional: true
  },
  'friends.$.a.b': {
    type: _SimpleSchema.SimpleSchema.Integer,
    optional: true
  },
  enemies: {
    type: Array
  },
  'enemies.$': {
    type: Object
  },
  'enemies.$.name': {
    type: String
  },
  'enemies.$.traits': {
    type: Array,
    optional: true
  },
  'enemies.$.traits.$': {
    type: Object
  },
  'enemies.$.traits.$.name': {
    type: String
  },
  'enemies.$.traits.$.weight': {
    type: Number
  }
});
var _default = friendsSchema;
exports.default = _default;
module.exports = exports.default;
module.exports.default = exports.default;
