// simpl-schema/dist/testHelpers/requiredSchema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _SimpleSchema = require("../SimpleSchema");

const requiredSchema = new _SimpleSchema.SimpleSchema({
  requiredString: {
    type: String
  },
  requiredBoolean: {
    type: Boolean
  },
  requiredNumber: {
    type: _SimpleSchema.SimpleSchema.Integer
  },
  requiredDate: {
    type: Date
  },
  requiredEmail: {
    type: String,
    regEx: _SimpleSchema.SimpleSchema.RegEx.Email
  },
  requiredUrl: {
    type: String,
    regEx: _SimpleSchema.SimpleSchema.RegEx.Url
  },
  requiredObject: {
    type: Object
  },
  'requiredObject.requiredNumber': {
    type: _SimpleSchema.SimpleSchema.Integer
  },
  optionalObject: {
    type: Object,
    optional: true
  },
  'optionalObject.requiredString': {
    type: String
  },
  anOptionalOne: {
    type: String,
    optional: true,
    min: 20
  }
});
requiredSchema.messageBox.messages({
  'regEx requiredEmail': '[label] is not a valid email address',
  'regEx requiredUrl': '[label] is not a valid URL'
});
var _default = requiredSchema;
exports.default = _default;
module.exports = exports.default;
module.exports.default = exports.default;
